Pumi2mfem field transfer updates [pumi2mfem_field_transfer]

mortezah opened the pull request:
## This PR
a) Cleans up of previous field transfer routines
b) Includes a new field transfer routine for Nedelec fields (which are now also implemented in PUMI) to be used for EM simulations.
c) Fixes a bug in the previous code. Previously the fact that vertex-to-tet connectivities can change during the conversion of the PUMI mesh to an MFEM mesh was ignored during field transfer routines. This is now fixed.   

## TODOs
* ~~pumi related tests on master branch fail because there is no example folder for pumi meshes~~ the required example meshes for theses tests to run are in the __additional large date repo__ here https://github.com/mfem/data
* ~~`make style` needs to be run on this~~ https://github.com/mortezah/mfem/commit/62bf39a011446ea1d930c62cbb2dd2604f41f653
* ~~more descriptive comments so that they show up in doxy~~ done in the last few commits.
<!--GHEX{"id":1454,"author":"mortezah","editor":"tzanio","reviewers":["dylan-copeland","psocratis"],"assignment":"2020-05-10T17:26:22-07:00","approval":"2020-05-25T17:45:19.306Z","merge":"2020-05-31T23:19:05.298Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1454](https://github.com/mfem/mfem/pull/1454) | @mortezah | @tzanio | @dylan-copeland + @psocratis | 05/10/20 | 05/25/20 | 05/31/20 | |
<!--ELBATXEHG-->

cwsmith commented on mesh/pumi.cpp:
```diff
@@ -923,223 +943,112 @@ void ParPumiMesh::UpdateMesh(const ParMesh* AdaptedpMesh)
+   // get downward vertices of PUMI element
+   apf::Downward vs;
+   int nv = apf_mesh->getDownward(tet,0,vs);
+   int pumi_vid[nv];
```
I think only the latest C++ standards, and some compiler extensions, will support variable length arrays.  To be safe side,  we should allocate the `pumi_vid` array or statically size it.

Some discussion here of VLAs in C++:
https://stackoverflow.com/questions/40633344/variable-length-arrays-in-c14

mortezah replied:
fixed in https://github.com/mortezah/mfem/commit/a23760dd801bc2f29245b1b0efc8c4fda43539fe

Also, ex6p runs cleanly under Valgrind.

dylan-copeland commented on mesh/pumi.hpp:
```diff
@@ -57,7 +56,7 @@ class PumiMesh : public Mesh
             bool fix_orientation = true);
 
-   using Mesh::Load;
+   /* using Mesh::Load; */
```
If this line is unnecessary, it should be deleted.

mortezah replied:
fixed in https://github.com/mortezah/mfem/commit/76b3f742c5f54234a1d284994fb08669366ffd7b

dylan-copeland commented:
I built the `master` version of https://github.com/SCOREC/core but this branch of MFEM gives the following build error:

mesh/pumi.cpp:41:9: error: no member named 'getElementNodeXis' in namespace 'apf'

Is there a particular version of `pumi` required?

psocratis commented:
> I built the `master` version of https://github.com/SCOREC/core but this branch of MFEM gives the following build error:
> 
> mesh/pumi.cpp:41:9: error: no member named 'getElementNodeXis' in namespace 'apf'
> 
> Is there a particular version of `pumi` required?

I'm getting the same thing

mortezah commented:
> I built the `master` version of https://github.com/SCOREC/core but this branch of MFEM gives the following build error:
> 
> mesh/pumi.cpp:41:9: error: no member named 'getElementNodeXis' in namespace 'apf'
> 
> Is there a particular version of `pumi` required?

@dylan-copeland  and @psocratis, thanks for pointing this out.

Those changes are in the `develop` branch of https://github.com/SCOREC/core. We have an automated system that periodically merges `develop` to `master`. I just realized that has not been happening lately. I will coordinate with @cwsmith for that merge. Meanwhile, you should be able to go through the build process by using the `develop` branch of core.

tzanio commented:
@mortezah, please then update the PUMI version requirement in `INSTALL`

mortezah commented:
> @mortezah, please then update the PUMI version requirement in `INSTALL`

Notes are added to `INSTALL` in https://github.com/mortezah/mfem/commit/b0935c7f63e109d2882a057067ee017245ec9bb3

mortezah commented:
@tzanio , I realized that one of the tests (ex25p) failed on Travis CI (MAC:Parallel) https://travis-ci.com/github/mfem/mfem/jobs/333525807

Are there specific requirements/instructions for MAC builds that I should follow?

psocratis commented:
> > @mortezah, please then update the PUMI version requirement in `INSTALL`
> 
> Notes were added to `INSTALL` in [mortezah@b0935c7](https://github.com/mortezah/mfem/commit/b0935c7f63e109d2882a057067ee017245ec9bb3)

We are able to run it now. Thanks

dylan-copeland commented:
I have a couple of suggestions to improve the master version of ex6p, which are not related to the changes made so far in this PR but should be added.

1) `make ex1p` should be `make ex6p`
2) A comment should be added, describing where to find the input files in the `data` repo. For example, see https://github.com/mfem/mfem/blob/master/miniapps/electromagnetics/joule.cpp#L89-L93

psocratis commented:
Hi @mortezah, 

I might be way off, but I think there is something going wrong here. When running ex6p the solution after the first refinement seems wrong, since the homogeneous essential boundary condition doesn't seem to be satisfied. This is what I get:

![image](https://user-images.githubusercontent.com/14336752/82399883-ed4fc580-9a0a-11ea-92ba-b3e855ed035e.png)

The solution from master for the same run looks like this

![image](https://user-images.githubusercontent.com/14336752/82399928-05bfe000-9a0b-11ea-9c34-8ad84fe69f58.png)

Then I compared the numbers from the solves, between the master and this branch and they are different except for the first solve. Here are the numbers: 

From master: 
AMR iteration 0
Number of unknowns: 22573
   Iteration :   0  (B r, r) = 0.05564 ...
   Iteration :  12  (B r, r) = 3.90428e-15
Average reduction factor = 0.283089
AMR iteration 1

Number of unknowns: 47477
   Iteration :   0  (B r, r) = 0.000408765 ...
   Iteration :  10  (B r, r) = 1.45049e-16
Average reduction factor = 0.238507

AMR iteration 2
Number of unknowns: 46162
   Iteration :   0  (B r, r) = 9.2109e-06 ...
   Iteration :  10  (B r, r) = 2.07973e-18
Average reduction factor = 0.233177


From this branch:

AMR iteration 0
Number of unknowns: 22573
   Iteration :   0  (B r, r) = 0.05564 ...
   Iteration :  12  (B r, r) = 3.90135e-15
Average reduction factor = 0.28308
glob_d_max = 0

AMR iteration 1
Number of unknowns: 47477
   Iteration :   0  (B r, r) = 0.202368 ...
   Iteration :  10  (B r, r) = 4.75066e-14
Average reduction factor = 0.233631

glob_d_max = 0

AMR iteration 2
Number of unknowns: 465784
   Iteration :   0  (B r, r) = 0.183108 ...
   Iteration :  10  (B r, r) = 1.34555e-13
Average reduction factor = 0.247349


So the picture suggests something is wrong in the ordering of dofs (maybe?) Notice that since the solutions from the second step are different, then the refinement is different and that results to a completely different 3rd mesh. 

Thanks, 

Socratis

mortezah commented:
Hi @psocratis 
Thanks for the comparison. I tried to replicate your results but I end up getting consistent results for the `master` and `pumi2mfem_field_transfer` branches of MFEM.

![master_vs_feature](https://user-images.githubusercontent.com/1325140/82408431-8dbcde80-9a39-11ea-926c-31fe58cb0521.png)

I am guessing you might not be using the most updated version of SCOREC/core code. Can you tell me the commit hash of the SCOREC/core used to produce your results above? 

I am using the latest commit on either the `develop` and/or the `master` branch of SCOREC/core for my results and they are both giving me consistent results.

psocratis commented:
Hi @mortezah 

Thanks for checking. I rerun everything with the latest commit from SCOREC/core and the latest commit from  `pumi2mfem_field_transfer`  and the results are the same as before. This happens both in debugging and optimized compilations. My setup is:

Apple clang version 11.0.0 (clang-1100.0.33.16) 
(Open MPI) 4.0.3

Socratis

psocratis commented on mesh/pumi.cpp:
```diff
@@ -1348,279 +1138,148 @@ void ParPumiMesh::FieldMFEMtoPUMI(apf::Mesh2* apf_mesh,
+   size_t elemNo = 0;
+   apf::MeshEntity* ent;
+   apf::MeshIterator* it = apf_mesh->begin(dim);
+   while ( ent = apf_mesh->iterate(it) )
```
Is this intended to be an assignment? If yes can you use double parenthesis so that the compiler warning is suppressed?

mortezah replied:
fixed in https://github.com/mortezah/mfem/commit/6f6e134d45b5a39f556cecfa284396cb515795f6

mortezah commented:
> Hi @mortezah
> 
> Thanks for checking. I rerun everything with the latest commit from SCOREC/core and the latest commit from `pumi2mfem_field_transfer` and the results are the same as before. This happens both in debugging and optimized compilations. My setup is:
> 
> Apple clang version 11.0.0 (clang-1100.0.33.16)
> (Open MPI) 4.0.3
> 
> Socratis

Sorry, I was not sure what you meant by the "results are the same as before". Did you mean you are still getting two very different outputs for the two branches?

One more question about your build: do you build with or without simmetrix?

psocratis commented:
> > Hi @mortezah
> > Thanks for checking. I rerun everything with the latest commit from SCOREC/core and the latest commit from `pumi2mfem_field_transfer` and the results are the same as before. This happens both in debugging and optimized compilations. My setup is:
> > Apple clang version 11.0.0 (clang-1100.0.33.16)
> > (Open MPI) 4.0.3
> > Socratis
> 
> @psocratis
> Sorry, I was not sure what you meant by the "results are the same as before". Did you mean you are still getting two very different outputs for the two branches?
> 
> One more question about your build: do you build with or without simmetrix?

@mortezah ,

Yes I still get the inconsistent results. I build without simmetrix. Is that required? 

Socratis

mortezah commented:
The reason I asked about simmetrix is that I usually build my stack with gcc/mpich/simmetrix and I am trying to build my stack using clang/openmpi to see if that is the cause of the problem and for that, I won't be able to use simmetrix. So I wanted to make sure my setup is as close as possible to yours.

Simmetrix is not required, and should not be the cause of this problem as far as I can tell. I will look more carefully into this and let you know.

mortezah commented:
Hi @psocratis 

Can you do me a favor? Where you PUMI libraries are installed there should be a `libma.a` file, can you run the following command (or its equivalent in mac if you are building on mac) and send me the output?

`nm -C libma.a | grep _rotation`

psocratis commented:
> Hi @psocratis
> 
> Can you do me a favor? Where your PUMI libraries are installed there should be a `libma.a` file, can you run the following command (or its equivalent in mac if you are building on mac) and send me the output?
> 
> `nm -C libma.a | grep _rotation`

Hi @mortezah , here is the output

`nm -C libma.a | grep _rotation
                 U ma::oct_rotation
                 U ma::tet_rotation
                 U ma::prism_rotation
                 U ma::pyramid_rotation
                 U ma::tet_inv_rotation
0000000000000530 S ma::oct_rotation
0000000000000060 S ma::tet_rotation
0000000000000420 S ma::prism_rotation
00000000000004b0 S ma::pyramid_rotation
0000000000000120 S ma::tet_inv_rotation
                 U ma::prism_rotation
                 U ma::pyramid_rotation`

psocratis commented:
@mortezah , 

Ok...this is embarrassing. It's my bad. 
For convenience  when I first built core i moved all the .a files in one directory /lib just to be easier to link from mfem. Somehow probably this messed up the paths or something (i don't know). I just modified the paths to pumi library to -L$(PUMI_DIR)/build/apf -lapf  -L$(PUMI_DIR)/build/pumi ... etc. and now it's working fine. 

So sorry for all the trouble. 

Socratis

psocratis commented:
> @mortezah ,
> 
> Ok...this is embarrassing. It's my bad.
> For convenience when I first built core i moved all the .a files in one directory /lib just to be easier to link from mfem. Somehow probably this messed up the paths or something (i don't know). I just modified the paths to pumi library to -L$(PUMI_DIR)/build/apf -lapf -L$(PUMI_DIR)/build/pumi ... etc. and now it's working fine.
> 
> So sorry for all the trouble.
> 
> Socratis

@mortezah,

actually it might not be the fact that I copied them, but maybe it's possible one or some of the .a files got corrupted or something between different pulls from the remote repo. In any case, I actually copied them again in one directory and linked again, and it's working fine now. 

Thanks, 

Socratis.

mortezah commented:
> tet_inv_rotation

Thanks for checking it. This looks OK. I was suspecting that `0000000000000120 S ma::tet_inv_rotation` would be missing but that is there! Hmmm!

Just to double-check, this libma.a is located where your mfem build looks for SCOREC libraries? The only thing that I could think of that might cause the problem is if your mfem libraries are built against an older version of SCOREC libraries, specifically an older version of libma.a.

mortezah commented:
> > @mortezah ,
> > Ok...this is embarrassing. It's my bad.
> > For convenience when I first built core i moved all the .a files in one directory /lib just to be easier to link from mfem. Somehow probably this messed up the paths or something (i don't know). I just modified the paths to pumi library to -L$(PUMI_DIR)/build/apf -lapf -L$(PUMI_DIR)/build/pumi ... etc. and now it's working fine.
> > So sorry for all the trouble.
> > Socratis
> 
> @mortezah,
> 
> actually it might not be the fact that I copied them, but maybe it's possible one or some of the .a files got corrupted or something between different pulls from the remote repo. In any case, I actually copied them again in one directory and linked again, and it's working fine now.
> 
> Thanks,
> 
> Socratis.

I think our comments crisscrossed. Thanks for double-checking.

psocratis commented on mesh/pumi.hpp:
```diff
@@ -57,8 +56,6 @@ class PumiMesh : public Mesh
    PumiMesh(apf::Mesh2* apf_mesh, int generate_edges = 0, int refine = 1,
             bool fix_orientation = true);
 
-   using Mesh::Load;
```
Clang gives an "overloaded virtual function" warning without this

mortezah replied:
Hi @psocratis 

Thanks for catching these. The two `getPumiNodeXis` routines are outdated and I will remove them shortly.

I am not able to get the clang warning on my build. Do you mind sharing your build options? I guess the one that I am looking for is the extra cxx flags that you are using. I guess using the MFEM build that would be specified by providing `CXXFLAGS` to `make config`

psocratis replied:
I am using the default debug flags for mfem. 
make pdebug -j

psocratis replied:
I think it's just `-g -Wall`

mortezah replied:
got it thanks.

mortezah replied:
These are fixed in the following two commits
https://github.com/mortezah/mfem/commit/78a8c79755344e40a85c0228c3744169ff9bd16a
https://github.com/mortezah/mfem/commit/40fb14767668b793de74b5f4b12025da26f12e2c

v-dobrev commented on INSTALL:
```diff
@@ -609,6 +611,7 @@ The specific libraries and their options are:
   URL: https://scorec.rpi.edu/pumi
+       https://github.com/SCOREC/core
   Options: PUMI_OPT, PUMI_LIB.
   Versions: PUMI >= 2.2.0.
```
I assume this version is no longer sufficient. Is there a released PUMI version that has the changes required for this PR?

mortezah replied:
The latest version of PUMI, as far as I know (@cwsmith  can comment on this), is 2.2.2 and that does not include these changes. I think the preferred way of using PUMI is to have the latest updates by pulling from either `master` or `develop` branch. I added a note in `INSTALL` here 
https://github.com/mortezah/mfem/blob/40fb14767668b793de74b5f4b12025da26f12e2c/INSTALL#L429 to that effect

tzanio replied:
This is not sufficient.

Please make sure the text after versions is correct, so users can indeed reproduce the build by themselves.

Maybe something like 

```
  Versions: PUMI >= 2.3.0
```

or

```
  Versions: Any version after commit xxxxxxx in PUMI's develop branch
```

mortezah replied:
I have updated `INSTALL` in the following commit

https://github.com/mortezah/mfem/commit/8d458616a77bc812c7d7626c519f9522ee9e3d38

cwsmith replied:
PUMI v2.2.3 now exists:
https://github.com/SCOREC/core/tree/v2.2.3
https://github.com/SCOREC/core/issues/308

tzanio replied:
In that case my preference will be to say `PUMI >= 2.3.0`

mortezah replied:
I will update that shortly.

mortezah replied:
I will update that shortly.

mortezah replied:
I pushed the change here https://github.com/mortezah/mfem/commit/a9bcd488cb99c88f76877e3840c1557e4ea1ad45

tzanio commented:
Merged in `next` for testing...

tzanio commented:
I merged this in `next` for testing, but please update `INSTALL` as requested above
